cardea/benchmark/benchmark.py

```diff
@@ -0,0 +1,365 @@
+    for i in range(task.run_num):
+        scores, models, hyperparameters = _evaluate_pipeline(i, pipeline, feature_matrix,
+                                                             task.pipeline_name,
+                                                             task.problem_name,
+                                                             task.dataset_name,
+                                                             task.beginning_stage,
+                                                             task.tuned, metrics)
+        results.append(scores)
+        records.append((models, hyperparameters))
+
+    # Store the output results.
+    if output_path is not None:
+        # Initialize the output directory.
+        if os.path.exists(output_path):
+            shutil.rmtree(output_path)
+        os.mkdir(output_path)
+
+        # Save task meta information
+        task.save_as(os.path.join(output_path, "meta.json"), "json")
+        matrix = 'F1 Macro'
+        best_index = np.argmax([scores[matrix] for scores in results])
+        models, hyperparameters = records[best_index]
+
+        # Save pipeline models if required
+        if save_model:
+            with open(os.path.join(output_path, "models.pkl"), 'wb') as f:
+                pickle.dump(models, f)
+
+        # Save pipeline hyperparameters if required
+        if save_hyperparameters and hyperparameters is not None:
+            with open(os.path.join(output_path, "hyperparameters.pkl"), 'wb') as f:
+                pickle.dump(hyperparameters, f)
+
+    return results
+
+
+def _evaluate_pipeline(run_id, pipeline, feature_matrix, pipeline_name=None, problem_name=None,
+                       dataset_name=None, beginning_stage=None, optimize=False, metrics=None):
+    """Evaluate a pipeline's performance on a target dataset with the given metrics.
+
+    Args:
+        run_id: int, the index to specify the execution to the pipeline.
+        pipeline: MLPipeline, a pipeline instance.
+        feature_matrix: pd.DataFrame, a dataframe consists of both feature values
+            and target values.
+        pipeline_name: str, the name of the pipeline.
+        problem_name: str, the name of the problem.
+        dataset_name: str, the name of the dataset.
+        beginning_stage: enumerate, the stage in which the benchmarking are applied, should be
+            either "data_loader", "problem_definition", "featurization".
+        optimize: boolean, whether to optimize the hyper-parameters of the pipeline.
+        metrics: dict, a dictionary in which metric functions are indexed by names.
+
+    Returns:
+        Pipeline evaluation results in a tuple: (performance, models, hyperparameters).
+    """
+    features, target = _split_features_target(feature_matrix, problem_name)
+
+    # TODO: digitize the labels for classifications before modeling.
+    if problem_name == 'LOS' and dataset_name == 'mimic-iii':
+        target = np.digitize(target, [0, 7])
```

Comment:
This part of the code is annoying, but I don't think we put this operation into the pipeline. To do so, we have to modify the modeler like `pipeline.predict(X, y, transform=(problem_name == 'LOS' and dataset_name == 'mimic-iii'))`, which will be even more annoying.

I'm thinking that this `digitalize` operation should be made in the featurization stage because the labels in classification problems are supposed to be concrete values from the beginning.

Reply:
Good point! I actually think the `problem definition` is the right component for this since featurization is just a wrapper. But yeah 100% agreed.